ntirpc/rpc/gss_internal.h

```diff
@@ -113,30 +112,17 @@ svc_rpc_gss_data *alloc_svc_rpc_gss_data(void)
 }
 
 static inline void
-unref_svc_rpc_gss_data(struct svc_rpc_gss_data *gd,
-		       uint32_t flags)
+unref_svc_rpc_gss_data(struct svc_rpc_gss_data *gd)
 {
-	u_int refcnt;
-	bool gd_locked = flags & SVC_RPC_GSS_FLAG_LOCKED;
-
-	refcnt = atomic_dec_uint32_t(&gd->refcnt);
+	mutex_lock(&gd->lock);
```

Comment:
If some code path fetches gd from hash table while some other thread trying to destroy is a bug by itself. It should never happen due to refcounting. The fundamental assertion I made is here that once the refcount goes to zero, it will never go up and no one will have the possession of the gd structure except the thread that made the refcnt zero. So there is no need to acquire gd->lock while destroying. I am NOT saying you can  eliminate the gd->lock. You don't need it while destroying, and it looks ugly to have it acquire in one function and release it in another function.